lib/child_process.js

```diff
@@ -527,29 +541,55 @@ function normalizeSpawnArguments(file, args, options) {
   }
 
   if (options.shell) {
-    const command = ArrayPrototypeJoin([file, ...args], ' ');
-    // Set the shell, switches, and commands.
-    if (process.platform === 'win32') {
-      if (typeof options.shell === 'string')
-        file = options.shell;
-      else
-        file = process.env.comspec || 'cmd.exe';
-      // '/d /s /c' is used only for cmd.exe.
-      if (RegExpPrototypeTest(/^(?:.*\\)?cmd(?:\.exe)?$/i, file)) {
-        args = ['/d', '/s', '/c', `"${command}"`];
-        windowsVerbatimArguments = true;
-      } else {
-        args = ['-c', command];
-      }
+    const command = file;
+    const commandArgs = ArrayPrototypeMap(args, String);
+    args = ['-c'];
+    let quote = quotePosixShArg;
+    let isCmd = false;
+    const testWindowsExe =
+      process.platform === 'win32' ?
+        (cmd, file) =>
+          RegExpPrototypeTest(
+            new RegExp(`^(?:.*\\\\)?${cmd}(?:\\.exe)?$`, 'i'),
+            file
+          ) :
+        (_c, _f) => false;
+    // Set the shell.
+    if (typeof options.shell === 'string') {
+      file = options.shell;
+    } else if (process.platform === 'win32') {
+      file = process.env.comspec || 'cmd.exe';
+    } else if (process.platform === 'android') {
+      file = '/system/bin/sh';
     } else {
-      if (typeof options.shell === 'string')
-        file = options.shell;
-      else if (process.platform === 'android')
-        file = '/system/bin/sh';
-      else
-        file = '/bin/sh';
-      args = ['-c', command];
+      file = '/bin/sh';
+    }
+
+    // Handle special args for special shells.
+    // '/d /s /c' is used only for cmd.exe.
+    if (testWindowsExe('cmd', file)) {
+      args = ['/d', '/s', '/c'];
+      quote = quoteCmdArg;
+      isCmd = windowsVerbatimArguments = true;
+    } else if (
+      testWindowsExe('(powershell|pwsh)', file) ||
+      StringPrototypeEndsWith(file, '/pwsh')
+    ) {
+      quote = quotePwshArg;
     }
+
+    // Make it opt-in... for now.
+    if (!options.shellEscape)
+      quote = (s) => s;
+    else if (typeof options.shellEscape == 'function')
+      quote = options.shellEscape;
+
+    // Cmd accepts quoted things verbatim under /s.
+    const commandArg = ArrayPrototypeJoin(
+      [command, ...ArrayPrototypeMap(commandArgs, quote)],
+      ' '
+    );
```

Comment:
We can avoid creating a new array here.
```suggestion
    const commandArg = commandArgs.length ?
      `${command} ${ArrayPrototypeJoin(ArrayPrototypeMap(commandArgs, quote), ' ')}`:
      command;
```